Create clhs points and sampleplan

gcperk opened the pull request:
This PR incorporates all the steps to 1) build the clhs latin hyper cube and then 2) create a full spatial sample plan and csv tracking sheet. There are a number of major functions within other subfunctions: from top down

1) build and assess clhs plans
- create_clhs (which uses kdaust/clhs package to build clhs points
- plot_clhs_cost (takes clhs reps and plots to determine least cost replicate)

2) Build the full sample plan and csv output
- create_sampleplan (which contains the subfunction (make_sampleplan (internal), build_site_transect (contains three internal subfunctions). 

I suspect these could be better assembled, to provide less requirements by the user, and more automations and could to with a high level suggestion when possible (@ateucher ).

ateucher commented on R/create_clhs.R:
```diff
@@ -0,0 +1,140 @@
+                        use.cpp = T,
+                        latlon = coords,
+                        min.dist = min_dist)
+        if(sum(templhs$final_obj_distance) == 0){
```
What does this mean when all of these equal zero? In my testing it never happened...

gcperk replied:
It means that the points are too close together to each other.

ateucher commented on R/create_clhs.R:
```diff
@@ -0,0 +1,140 @@
+  } else {
+    cli::cli_alert_success("Gen-R-ating multiple slices...")
+
+    for (snum in 1:num_slices) {
```
```suggestion
    for (snum in seq_len(num_slices)) {
```

ateucher commented on R/create_clhs.R:
```diff
@@ -0,0 +1,140 @@
+                        simple = FALSE,
+                        progress = TRUE,
+                        cost= "cost",
+                        use.cpp = T,
```
```suggestion
                        use.cpp = TRUE,
```

ateucher commented on R/create_clhs.R:
```diff
@@ -0,0 +1,140 @@
+    cli::cli_alert_warning("Some points fall within minimum distance!")}
+
+  out <- as.data.frame(samp_dat[templhs$index_samples, ])
+  out$slice_num <- rep(num_slices:1, each = n_points)
+  out$point_num <- rep(1:n_points, times = num_slices)
```
```suggestion
  out$slice_num <- rep(rev(seq_len(num_slices)), each = n_points)
  out$point_num <- rep(seq_len(n_points), times = num_slices)
```

ateucher commented on R/create_sampleplan.R:
```diff
@@ -0,0 +1,72 @@
+}
+
+
+make_sampleplan <- function(clhs_set) {
```
`make_sample_plan()` needs `clhs_dir`, `mask_dir`, `cost`, and `outdir` parameters, passed from the outer function...

gcperk replied:
I think I added these now, so fingers crossed it work?

ateucher commented on R/create_sampleplan.R:
```diff
@@ -0,0 +1,72 @@
+
+make_sampleplan <- function(clhs_set) {
+  boi <- stringr::str_extract(clhs_set, "[^_]+")
+  sample_points <- sf::st_read(fs::path(clhs_dir, pattern = clhs_set), quiet = T)
```
```suggestion
  sample_points <- sf::st_read(fs::path(clhs_dir, pattern = clhs_set), quiet = TRUE)
```

ateucher commented on R/create_sampleplan.R:
```diff
@@ -0,0 +1,72 @@
+  boi <- stringr::str_extract(clhs_set, "[^_]+")
+  sample_points <- sf::st_read(fs::path(clhs_dir, pattern = clhs_set), quiet = T)
+
+  mask_poly <- sf::st_read(fs::path(mask_dir, pattern = paste0(boi, "_exclude_poly.gpkg")), quiet = T)
```
```suggestion
  mask_poly <- sf::st_read(fs::path(mask_dir, pattern = paste0(boi, "_exclude_poly.gpkg")), quiet = TRUE)
```

ateucher commented on R/create_sampleplan.R:
```diff
@@ -0,0 +1,72 @@
+
+  cost <- PEMprepr:::read_spatrast_if_necessary(cost)
+
+  purrr::map(clhs_set, make_sampleplan)
```
```suggestion
  purrr::map(clhs_set, \(x) make_sampleplan(x, clhs_dir, mask_dir, cost, out_dir))
```

ateucher replied:
See also https://github.com/ninoxconsulting/PEMsamplr/pull/7/files#r1884684283

ateucher commented on R/create_sampleplan.R:
```diff
@@ -0,0 +1,72 @@
+}
+
+
+make_sampleplan <- function(clhs_set) {
```
```suggestion
make_sampleplan <- function(clhs_set, clhs_dir, mask_dir, cost, out_dir) {
```

ateucher commented on R/create_cost_penalty.R:
```diff
@@ -98,7 +98,11 @@ create_cost_penalty <- function(vec_dir = fs::path(PEMprepr::read_fid()$dir_1010
-    rvri12_class <- .assign_highcost(file.path(vec_dir, "vri_class1_2.gpkg"), costval = costval, cost = cost)
+    rvri12_class <- .assign_highcost(
+      file.path(vec_dir, "vri_class1_2.gpkg"),
+      costval = costval, # Should cosval here be vri_cost? Docs suggest it is for class 1 and 2
```
Should costval here be vri_cost? Docs suggest it is for class 1 and 2

gcperk replied:
No, there are two levels of costs and higher cost goes to VRI classe 1 and 2 and slightly less for VRI class 3...  I know a bit confusing here! See line 117.

ateucher commented on R/create_cost_penalty.R:
```diff
@@ -110,11 +114,13 @@ create_cost_penalty <- function(vec_dir = fs::path(PEMprepr::read_fid()$dir_1010
-    rvri3_class <- .assign_highcost(file.path(vec_dir, "vri_class3.gpkg"), costval = vri_cost, cost = cost)
+    rvri3_class <- .assign_highcost(
+      file.path(vec_dir, "vri_class3.gpkg"),
+      costval = vri_cost, # Should costval = costval here, instead of vri_cost?
```
Should costval = costval here, instead of vri_cost?

ateucher commented on R/create_cost_penalty.R:
```diff
@@ -243,10 +241,23 @@ create_cost_penalty <- function(vec_dir = fs::path(PEMprepr::read_fid()$dir_1010
   return(rhc)
 }
+
+make_steep_slopes <- function(dem, maxval, costval) {
```
I pulled this bit out into its own function

ateucher reviewed (commented):
Hi Gen! I haven't totally finished, but I think this is a reasonable place to pause, in case you have time to work on my feedback so far. I have been working through the PEMr vignette as I go through the functions. I am just starting to look at `create_sample_plan()`, `make_sampleplan()`, and `build_site_transect()`.

I pushed up a few small changes I made, so it would be great if you could have a look at those

ateucher commented on R/build_site_transects.R:
```diff
@@ -0,0 +1,209 @@
+  cli::cli_alert_success("generating site points")
+
+
+  for (i in 1:nrow(sample_points_clhs)) {
+    # i = 1
+    pnt <- sample_points_clhs[i, ]
+    pGeom <- sf::st_geometry(pnt)
+    pGeom <- pGeom + c(0, centroid_distance)
+    pnt_feat <- sf::st_set_geometry(pnt, pGeom)
+
+    rotated_points <- sf::st_sf(sf::st_sfc()) |> sf::st_set_crs(3005)
+
+    rotated_points <- do.call(rbind, lapply(rotation_angles, function(Bear) {
+      Feature_geo <- sf::st_geometry(pnt_feat)
+      PivotPoint <- sf::st_geometry(pnt)
+      d <- ifelse(Bear > 180, pi * ((Bear - 360) / 180), pi * (Bear / 180))
+      rFeature <- (Feature_geo - PivotPoint) * .rot(d) + PivotPoint
+      rFeature <- sf::st_set_crs(rFeature, sf::st_crs(pnt_feat))
+      pnt_feat$geometry <- sf::st_geometry(rFeature)
+      pnt_feat$Rotation <- Bear
+      pnt_feat <- pnt_feat |> sf::st_set_crs(3005)
+      pnt_feat
+    }))
+
+    sample_points_rotations <- rbind(rotated_points, sample_points_rotations)
+  }
```
Also, the map call that makes `rotated_points` - can we use `.rotFeature()` here instead of the anonymous function? They seem almost identical...

gcperk replied:
Nice point @ateucher . They are doing two different things, but you are correct the function looks very similar. The points are being generated in a circle around the clhs point, while each triangle is being rotated on a random drawn axis around its center point. I tried updating the point generation but it is breaking my brain so I will add as improvement in issues to fix/review.

ateucher commented on R/build_site_transects.R:
The `.by` argument is nice because it avoids having to ungroup, and it avoids non-standard evaluation shenanigans

```suggestion
    dplyr::summarise(.by = "id") |>
```

ateucher commented on R/create_sampleplan.R:
```diff
@@ -0,0 +1,72 @@
+  cli::cat_line()
+  cli::cli_alert_success("sample plan generated for {.var clhs_set}")
+
+  allpoints <- grep("points_all", sf::st_layers(file.path(out_dir, "s1_sampling.gpkg"))$name, value = T)
```
```suggestion
  allpoints <- grep("points_all", sf::st_layers(file.path(out_dir, "s1_sampling.gpkg"))$name, value = TRUE)
```

gcperk replied:
Updated all

ateucher reviewed (approved):
Hey Gen,

Nice work on this - it's pretty cool the amount that these functions do, very impressive!

I didn't really catch any major issues, with the exception of needing to call `make_sample_plan()` with the necessary arguments. 

Also, some rlang/dplyr stuff to make the non-standard evaluation work as expected. To that end, I did run `usethis::use_import_from("rlang", ".data"), and pushed that, which adds package-level documentation and also allows the use of the `.data` pronoun in dplyr calls.

gcperk commented on R/create_clhs.R:
```diff
@@ -0,0 +1,140 @@
+  if (num_slices < 1) {
+    cli::cli_abort("Hold up! {.var num_slices} must have at least one slice.")
+  }
+
```
I have a note in the placeholder script to generate the cost layer to ensure the name is "cost", but I will add a check to make sure the names is cost and numeric.

gcperk commented on R/create_clhs.R:
```diff
@@ -0,0 +1,140 @@
+                                xy = TRUE,
+                                as.df = F
+  )
+  samp_dat <- samp_dat[!is.na(samp_dat[, "cost"]) & !is.infinite(samp_dat[, ncol(samp_dat)]), ]
```
@ateucher - yes this is specifically looking at the subsample of points which the cost is NA or infinite. The last , is to define to keep all columns. I dont want to use the drop = FALSE here as we dont want to put clhs points where there are no costs so these rows (pixals) are removed from the sample drawn.

gcperk replied:
updated to 'samp_dat <- samp_dat[!is.na(samp_dat[, "cost"]) & !is.infinite(samp_dat[, "cost"]), ]'

ateucher replied:
@gcperk just to clarify - `drop = FALSE` within `[]` doesn't have to do with droopping rows. It prevents the object from "dropping" from a data.frame to a vector if the selection ends up with only one column... eg.:

``` r
cars[cars$speed > 20, "speed"]
#> [1] 22 23 24 24 24 24 25

cars[cars$speed > 20, "speed", drop = FALSE]
#>    speed
#> 44    22
#> 45    23
#> 46    24
#> 47    24
#> 48    24
#> 49    24
#> 50    25
```

<sup>Created on 2024-12-18 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

gcperk replied:
Mind blown! Thanks for the clarrification!

gcperk commented on R/create_clhs.R:
```diff
@@ -0,0 +1,140 @@
+#' @param all_cov A `spatRast` stack with all landscape covariates and masked
+#' cost layer. Note all layers excluding the cost layer should be factors.
+#' @param num_slices A `numeric` representing the number of slices to be built
+#' @param to_include A `sf`dataframe of already sample points to include in plan
```
Yes..... urgh. I might need to do some more testing on this one so I have added this as an other issue

gcperk commented on R/create_clhs.R:
```diff
@@ -0,0 +1,140 @@
+    size <- n_points
+  } else {
+    inc_pts <- terra::extract(all_cov, to_include)
+    inc_pts <- inc_pts[, -(1)]
```
added to #9

gcperk commented on R/build_site_transects.R:
```diff
@@ -0,0 +1,209 @@
+
+  # create paired outputs
+  sample_points_clhs <- sf::st_as_sf(sample_points) |> sf::st_transform(3005)
+  sample_points_clhs$aoi <- NA
```
Added

gcperk commented on R/build_site_transects.R:
```diff
@@ -0,0 +1,209 @@
+    # i = 1
+    pnt <- sample_points_clhs[i, ]
+    pGeom <- sf::st_geometry(pnt)
+    pGeom <- pGeom + c(0, centroid_distance)
```
I thought it was best practice to avoid comments within functions? I can see it if fairly cryptic.  Unlikely to be more than 400, as this is the min distance apart so that paired triangles do not overlap (no matter the rotation). Although might be helpful at some point... Im concerted this param is so far buried the user will need to have many many params to run the sample plan? thoughts @ateucher ?

gcperk commented on R/build_site_transects.R:
```diff
@@ -0,0 +1,209 @@
+      is.na(cost) ~ FALSE,
+      TRUE ~ TRUE
+    )) |>
+    dplyr::select(-cost)
```
added

gcperk commented on R/build_site_transects.R:
updated all l

gcperk commented on R/build_site_transects.R:
```diff
@@ -0,0 +1,209 @@
+  cost <- terra::extract(cost, sample_points_rotations, ID = FALSE)
+  sample_points_rotations <- cbind(sample_points_rotations, cost)
+
+  sample_points_low_cost <- do.call(rbind, lapply(split(
+    sample_points_rotations,
+    list(
+      sample_points_rotations$slice_num,
+      sample_points_rotations$point_num
+    )
+  ), function(df) {
+    df <- subset(df, aoi == TRUE)
+    df[which.min(df$cost), ]
+  }))
+
+  sample_points_low_cost <- sample_points_low_cost |>
+    dplyr::select(-c("cost", "Rotation", "aoi"))
```
added

gcperk commented on R/build_site_transects.R:
updated

gcperk commented on R/build_site_transects.R:
updated

gcperk commented on R/build_site_transects.R:
updated

gcperk commented on R/build_site_transects.R:
as above

gcperk commented on R/build_site_transects.R:
updated

gcperk commented on R/build_site_transects.R:
```diff
@@ -0,0 +1,209 @@
+    sample_points_rotations <- rbind(rotated_points, sample_points_rotations)
+  }
+
+  sample_points_rotations <- sf::st_as_sf(sample_points_rotations, crs = 3005) |>
```
added

gcperk commented:
@ateucher - thanks for the thorough review. I updated and added most comments. I really like the addition of Purr for rbind as this seems much faster ( I was using do.call to avoid non-standard eval issues, but thanks for the fix!). There are a few components that need some more detailed review so I have added issues (#10 , #9 , #8 ) to keep these on the radar.
